File: agent-factory/self_improve.py
"""self_improve.py — bounded, logged self-tuning.

After each batch the factory reviews its own performance (success rate, cost,
latency) and adjusts a FIXED set of numeric knobs in config.json, each clamped
to declared bounds (SANDBOX.md rule 4). It never edits code. Every decision is
appended to LEARNINGS.md and to config.tuning_history.

The knobs it may turn:
  - concurrency            (how many workers run at once)
  - routing.complexity_threshold  (how eager it is to escalate to the strong model)
  - poll_interval_seconds  (base scheduler cadence)
"""
from __future__ import annotations

import core
import task_queue


def review_and_tune(batch: int) -> dict:
    cfg = core.load_config()
    stats = task_queue.batch_stats(batch)
    targets = cfg["targets"]
    bounds = cfg["bounds"]

    before = {
        "concurrency": cfg["concurrency"],
        "complexity_threshold": cfg["routing"]["complexity_threshold"],
        "poll_interval_seconds": cfg["poll_interval_seconds"],
    }
    reasons = []

    # queue pressure right now
    pending = task_queue.counts()["pending"]

    # --- concurrency: scale with success + remaining backlog -----------------
    if stats["success_rate"] >= targets["min_success_rate"] and pending > cfg["concurrency"]:
        new_conc = core.clamp(cfg["concurrency"] + 1, *bounds["concurrency"])
        if new_conc != cfg["concurrency"]:
            reasons.append(f"healthy success {stats['success_rate']:.0%} with "
                           f"{pending} pending -> concurrency {cfg['concurrency']}→{new_conc}")
            cfg["concurrency"] = new_conc
    elif stats["success_rate"] < targets["min_success_rate"]:
        new_conc = core.clamp(cfg["concurrency"] - 1, *bounds["concurrency"])
        if new_conc != cfg["concurrency"]:
            reasons.append(f"degraded success {stats['success_rate']:.0%} -> "
                           f"back off concurrency {cfg['concurrency']}→{new_conc}")
            cfg["concurrency"] = new_conc

    # --- routing threshold: defend the avg-cost target -----------------------
    th = cfg["routing"]["complexity_threshold"]
    if stats["avg_cost_usd"] > targets["max_avg_cost_usd"]:
        new_th = round(core.clamp(th + 0.05, *bounds["complexity_threshold"]), 3)
        if new_th != th:
            reasons.append(f"avg cost ${stats['avg_cost_usd']:.4f} over target "
                           f"${targets['max_avg_cost_usd']:.4f} -> raise escalation "
                           f"bar {th}→{new_th} (escalate less)")
            cfg["routing"]["complexity_threshold"] = new_th
    elif stats["avg_cost_usd"] < targets["max_avg_cost_usd"] * 0.5 and stats["completed"]:
        # plenty of cost headroom — allow slightly more escalation for quality
        new_th = round(core.clamp(th - 0.03, *bounds["complexity_threshold"]), 3)
        if new_th != th:
            reasons.append(f"avg cost ${stats['avg_cost_usd']:.4f} well under target "
                           f"-> lower escalation bar {th}→{new_th} (allow more quality)")
            cfg["routing"]["complexity_threshold"] = new_th

    # --- cadence: poll faster when backlog remains ---------------------------
    pi = cfg["poll_interval_seconds"]
    if pending >= cfg["concurrency"] * 2:
        new_pi = core.clamp(pi - 1, *bounds["poll_interval_seconds"])
        if new_pi != pi:
            reasons.append(f"deep backlog ({pending}) -> poll faster {pi}→{new_pi}s")
            cfg["poll_interval_seconds"] = new_pi
    elif pending == 0:
        new_pi = core.clamp(pi + 1, *bounds["poll_interval_seconds"])
        if new_pi != pi:
            reasons.append(f"queue drained -> poll slower {pi}→{new_pi}s")
            cfg["poll_interval_seconds"] = new_pi

    after = {
        "concurrency": cfg["concurrency"],
        "complexity_threshold": cfg["routing"]["complexity_threshold"],
        "poll_interval_seconds": cfg["poll_interval_seconds"],
    }

    entry = {
        "ts": core.now_iso(), "batch": batch, "stats": stats,
        "pending_after": pending, "before": before, "after": after,
        "reasons": reasons or ["no change — metrics within targets"],
    }
    cfg.setdefault("tuning_history", []).append(entry)
    core.save_config(cfg)
    _append_learnings(entry)
    return entry


def _append_learnings(entry: dict) -> None:
    s = entry["stats"]
    lines = [
        f"\n## Batch {entry['batch']} — {entry['ts']}",
        f"- Completed **{s['completed']}**, failed {s['failed']}, "
        f"success **{s['success_rate']:.0%}**",
        f"- Spend ${s['total_cost_usd']:.5f} (avg ${s['avg_cost_usd']:.5f}/task), "
        f"avg latency {s['avg_latency_ms']:.0f}ms",
        f"- Model mix: {s['model_mix'] or '{}'}",
        f"- Pending after batch: {entry['pending_after']}",
        "- Adjustments:",
    ]
    for r in entry["reasons"]:
        lines.append(f"  - {r}")
    with open(core.LEARNINGS, "a", encoding="utf-8") as fh:
        fh.write("\n".join(lines) + "\n")

File: agent-factory/src/selfimprove.js
// Bounded self-improvement. After each batch it compares measured performance to
// targets and adjusts ONLY config/factory-config.json values, each clamped to the
// declared _bounds. It never edits core source. Findings go to LEARNINGS.md.
import { saveConfig, ts } from './util.js';
import { appendInside } from './safepath.js';

function getBound(cfg, dotted) {
  return cfg._bounds[dotted];
}
function clamp(v, b) { return b ? Math.max(b.min, Math.min(b.max, v)) : v; }
function getNested(o, dotted) { return dotted.split('.').reduce((a, k) => a?.[k], o); }
function setNested(o, dotted, v) {
  const ks = dotted.split('.'); const last = ks.pop();
  ks.reduce((a, k) => a[k], o)[last] = v;
}

export function improve({ cfg, stats, queueDepth }) {
  const changes = [];
  const findings = [];
  const t = cfg.targets;

  const successRate = stats.processed ? stats.succeeded / stats.processed : 1;
  const avgCost = stats.processed ? stats.cost_usd / stats.processed : 0;
  const avgLatency = stats.avg_latency_ms || 0;

  const adjust = (dotted, newVal, why) => {
    const b = getBound(cfg, dotted);
    const clamped = clamp(newVal, b);
    const old = getNested(cfg, dotted);
    if (clamped !== old) {
      setNested(cfg, dotted, clamped);
      changes.push({ param: dotted, from: old, to: clamped, why });
    }
  };

  // 1. Reliability first: low success -> fewer concurrent workers + escalate sooner.
  if (successRate < t.min_success_rate) {
    findings.push(`Success ${(successRate*100).toFixed(0)}% < target ${(t.min_success_rate*100)}%; reducing concurrency and escalating models sooner.`);
    adjust('concurrency', cfg.concurrency - 1, 'low success rate');
    adjust('routing.escalate_to_tier2_at', cfg.routing.escalate_to_tier2_at - 0.05, 'use stronger models sooner');
  }

  // 2. Cost control: over budget -> prefer cheaper models (raise escalate thresholds).
  if (avgCost > t.max_avg_cost_usd) {
    findings.push(`Avg cost $${avgCost.toFixed(4)} > target $${t.max_avg_cost_usd}; raising escalation thresholds to favor cheaper tiers.`);
    adjust('routing.escalate_to_tier2_at', cfg.routing.escalate_to_tier2_at + 0.05, 'over cost budget');
    adjust('routing.escalate_to_tier3_at', cfg.routing.escalate_to_tier3_at + 0.05, 'over cost budget');
  }

  // 3. Latency / throughput: slow batch but healthy success -> add a worker.
  if (avgLatency > t.max_avg_latency_ms && successRate >= t.min_success_rate) {
    findings.push(`Avg latency ${avgLatency.toFixed(0)}ms > target ${t.max_avg_latency_ms}ms; adding a worker for parallelism.`);
    adjust('concurrency', cfg.concurrency + 1, 'reduce wall-clock latency');
  }

  // 4. Healthy + backlog -> scale up throughput.
  if (successRate >= t.min_success_rate && avgCost <= t.max_avg_cost_usd && queueDepth > cfg.batch_size) {
    findings.push(`Healthy with backlog of ${queueDepth}; scaling batch size and concurrency up for throughput.`);
    adjust('batch_size', cfg.batch_size + 2, 'drain backlog faster');
    adjust('concurrency', cfg.concurrency + 1, 'drain backlog faster');
  }

  // 5. Healthy + empty queue + over-provisioned -> tighten cadence (cheaper idling).
  if (queueDepth === 0 && cfg.concurrency > 2) {
    findings.push('Queue empty; trimming concurrency back toward baseline.');
    adjust('concurrency', cfg.concurrency - 1, 'no backlog');
  }

  if (findings.length === 0) findings.push('Within all targets; no parameter changes.');

  cfg.version += 1;
  saveConfig(cfg);

  // Append a dated entry to LEARNINGS.md
  const entry =
    `\n## ${ts()} — batch #${stats.n}\n` +
    `- processed: ${stats.processed} · succeeded: ${stats.succeeded} · failed: ${stats.failed}\n` +
    `- success rate: ${(successRate*100).toFixed(1)}% · avg cost: $${avgCost.toFixed(4)} · avg latency: ${avgLatency.toFixed(0)}ms · batch cost: $${stats.cost_usd.toFixed(4)}\n` +
    `- queue depth after: ${queueDepth}\n` +
    findings.map(f => `- finding: ${f}`).join('\n') + '\n' +
    (changes.length
      ? changes.map(c => `- tuned **${c.param}**: ${c.from} → ${c.to} (${c.why})`).join('\n') + '\n'
      : '- no config changes\n');
  appendInside('LEARNINGS.md', entry);

  return { changes, findings, metrics: { successRate, avgCost, avgLatency } };
}
